python/src/communication_interfaces_bindings.cpp

```diff
@@ -0,0 +1,60 @@
+#include "communication_interfaces_bindings.hpp"
+
+#include <communication_interfaces/exceptions/SocketConfigurationException.hpp>
+#include <communication_interfaces/sockets/ISocket.hpp>
+
+#define STRINGIFY(x) #x
+#define MACRO_STRINGIFY(x) STRINGIFY(x)
+
+using namespace communication_interfaces;
+
+class PySocket : public sockets::ISocket, public std::enable_shared_from_this<PySocket> {
+public:
+  using sockets::ISocket::ISocket;
+
+  void open() override { PYBIND11_OVERRIDE_PURE(void, ISocket, open, ); }
+
+  bool receive_bytes(std::string& buffer) override { PYBIND11_OVERRIDE_PURE(bool, ISocket, receive_bytes, buffer); }
+
+  bool send_bytes(const std::string& buffer) override { PYBIND11_OVERRIDE_PURE(bool, ISocket, send_bytes, buffer); }
+
+  void close() override { PYBIND11_OVERRIDE(void, ISocket, close, ); }
+};
+
+PYBIND11_MODULE(communication_interfaces, m) {
+  m.doc() = "Python bindings for communication interfaces";
+
+#ifdef MODULE_VERSION_INFO
+  m.attr("__version__") = MACRO_STRINGIFY(MODULE_VERSION_INFO);
+#else
+  m.attr("__version__") = "dev";
+#endif
+
+  auto m_sub_ex = m.def_submodule("exceptions", "Submodule for custom communication interfaces exceptions");
+  py::register_exception<exceptions::SocketConfigurationException>(
+      m_sub_ex, "SocketConfigurationError", PyExc_RuntimeError);
+
+  auto m_sub_sock = m.def_submodule("sockets", "Submodule for communication interfaces sockets");
+
+  py::class_<sockets::ISocket, std::shared_ptr<sockets::ISocket>, PySocket>(m_sub_sock, "ISocket")
+      .def("open", &sockets::ISocket::open, "Perform configuration steps to open the socket for communication")
+      .def(
+          "receive_bytes",
+          [](sockets::ISocket& socket) -> py::object {
+            std::string buffer;
+            auto res = socket.receive_bytes(buffer);
+            if (res) {
+              buffer.erase(std::find(buffer.begin(), buffer.end(), '\0'), buffer.end());
+              return py::bytes(buffer);
```

Comment:
comment: here the pattern changes a bit compared C++ implementation.

Instead of modifying the buffer by reference, it returns the received bytes (or None). This is quite nice to use, but just worth noting that it a different signature. 

And, we ended up using `std::string` for the send / receive buffer in C++, but you are returning `py::bytes` here. Do you think bytes are preferable? If you return the std::string buffer directly it will come through as a utf-8 decoded python string and handle the null termination automatically.

Reply:
I thought that we ended up using `std::string` because that could contain anything, even bytes. If we send bytes (for example an encoded state) and just decode it as utf8 string, it's not going to work, hence I chose bytes.

I went for [this](https://pybind11.readthedocs.io/en/stable/advanced/cast/strings.html#return-c-strings-without-conversion) option at the end and I think it's quite nice to use because we are sure that there is no hidden conversion in the message that we receive.

Reply:
gotcha, happy to keep with `py::bytes` then. for the other point, are you happy with the signature being different? for me it's fine but I just wanted to comment on it.

Reply:
Yeah I think it's more pythonic this way